Mathlib/Topology/Category/Profinite/Limits.lean

```diff
@@ -0,0 +1,211 @@
+
+open CategoryTheory
+
+section Pullbacks
+
+variable {X Y B : Profinite.{u}} (f : X ⟶ B) (g : Y ⟶ B)
+
+/--
+The pullback of two morphisms `f, g` in `Profinite`, constructed explicitly as the set of
+pairs `(x, y)` such that `f x = g y`, with the topology induced by the product.
+-/
+def pullback : Profinite.{u} :=
+  letI set := { xy : X × Y | f xy.fst = g xy.snd }
+  haveI : CompactSpace set := isCompact_iff_compactSpace.mp
+    (isClosed_eq (f.continuous.comp continuous_fst) (g.continuous.comp continuous_snd)).isCompact
+  Profinite.of set
+
+/-- The projection from the pullback to the first component. -/
+def pullback.fst : pullback f g ⟶ X where
+  toFun := fun ⟨⟨x, _⟩, _⟩ => x
+  continuous_toFun := Continuous.comp continuous_fst continuous_subtype_val
+
+/-- The projection from the pullback to the second component. -/
+def pullback.snd : pullback f g ⟶ Y where
+  toFun := fun ⟨⟨_, y⟩, _⟩ => y
+  continuous_toFun := Continuous.comp continuous_snd continuous_subtype_val
+
+@[reassoc]
+lemma pullback.condition : pullback.fst f g ≫ f = pullback.snd f g ≫ g := by
+  ext ⟨_, h⟩
+  exact h
+
+/--
+Construct a morphism to the explicit pullback given morphisms to the factors
+which are compatible with the maps to the base.
+This is essentially the universal property of the pullback.
+-/
+def pullback.lift {Z : Profinite.{u}} (a : Z ⟶ X) (b : Z ⟶ Y) (w : a ≫ f = b ≫ g) :
+    Z ⟶ pullback f g where
+  toFun := fun z => ⟨⟨a z, b z⟩, by apply_fun (· z) at w; exact w⟩
+  continuous_toFun := by
+    apply Continuous.subtype_mk
+    rw [continuous_prod_mk]
+    exact ⟨a.continuous, b.continuous⟩
+
+@[reassoc (attr := simp)]
+lemma pullback.lift_fst {Z : Profinite.{u}} (a : Z ⟶ X) (b : Z ⟶ Y) (w : a ≫ f = b ≫ g) :
+  pullback.lift f g a b w ≫ pullback.fst f g = a := rfl
+
+@[reassoc (attr := simp)]
+lemma pullback.lift_snd {Z : Profinite.{u}} (a : Z ⟶ X) (b : Z ⟶ Y) (w : a ≫ f = b ≫ g) :
+  pullback.lift f g a b w ≫ pullback.snd f g = b := rfl
+
+lemma pullback.hom_ext {Z : Profinite.{u}} (a b : Z ⟶ pullback f g)
+    (hfst : a ≫ pullback.fst f g = b ≫ pullback.fst f g)
+    (hsnd : a ≫ pullback.snd f g = b ≫ pullback.snd f g) : a = b := by
+  ext z
+  apply_fun (· z) at hfst hsnd
+  apply Subtype.ext
+  apply Prod.ext
+  · exact hfst
+  · exact hsnd
+
+/-- The pullback cone whose cone point is the explicit pullback. -/
+@[simps! pt π]
+def pullback.cone : Limits.PullbackCone f g :=
+  Limits.PullbackCone.mk (pullback.fst f g) (pullback.snd f g) (pullback.condition f g)
+
+/-- The explicit pullback cone is a limit cone. -/
+@[simps! lift]
+def pullback.isLimit : Limits.IsLimit (pullback.cone f g) :=
+  Limits.PullbackCone.isLimitAux _
+    (fun s => pullback.lift f g s.fst s.snd s.condition)
+    (fun _ => pullback.lift_fst _ _ _ _ _)
+    (fun _ => pullback.lift_snd _ _ _ _ _)
+    (fun _ _ hm => pullback.hom_ext _ _ _ _ (hm .left) (hm .right))
+
+section Isos
+
+noncomputable
+def pullbackIsoPullback : Profinite.pullback f g ≅ Limits.pullback f g :=
```

Comment:
There is a missing docstring here (which CI does not catch since the corresponding linter is still disabled in CI).